src/error-table.rkt

```diff
@@ -1,70 +1,157 @@
 #lang racket
 
-(require racket/set math/bigfloat)
+(require racket/set math/bigfloat racket/hash)
 (require "points.rkt" "syntax/types.rkt" "core/localize.rkt" "common.rkt"
-         "ground-truth.rkt" "syntax/sugar.rkt" "timeline.rkt" "programs.rkt")
+         "ground-truth.rkt" "syntax/sugar.rkt" "alternative.rkt" "programs.rkt"
+         "float.rkt" "config.rkt")
 
-(provide actual-errors predicted-errors)
+(provide actual-errors predicted-errors make-flow-table calculate-confusion
+         calculate-confusion-maybe)
+
+(define *top-3* (make-parameter #f))
+
+(define (take-top-n lst)
+  (if (*top-3*)
+      (take-n 3 lst)
+      lst))
+
+(define (take-n n lst)
+  (match lst
+    ['() '()]
+    [(cons x xs)
+     (if (= n 0)
+         '()
+         (cons x (take-n (- n 1) xs)))]))
+
+(define (constant? expr)
+  (cond
+    [(list? expr) (andmap constant? (rest expr))]
+    [(symbol? expr) #f]
+    [else #t]))
 
 (define (actual-errors expr pcontext)
   (match-define (cons subexprs pt-errorss)
     (flip-lists
      (hash->list (first (compute-local-errors (list (all-subexpressions expr))
                                        (*context*))))))
 
-
   (define pt-worst-subexpr
     (append* (reap [sow]
                    (for ([pt-errors (in-list pt-errorss)]
-                              [(pt _) (in-pcontext pcontext)])
+                         [(pt _) (in-pcontext pcontext)])
                      (define sub-error (map cons subexprs pt-errors))
                      (define filtered-sub-error
                        (filter (lambda (p) (> (cdr p) 16)) sub-error))
                      (define mapped-sub-error
                        (map (lambda (p) (cons (car p) pt))
                             filtered-sub-error))
-                      (unless (empty? mapped-sub-error)
-                        (sow mapped-sub-error))))))
-  
+                     (unless (empty? mapped-sub-error)
+                       (sow mapped-sub-error))))))
+
   (for/hash ([group (in-list (group-by car pt-worst-subexpr))])
     (let ([key (caar group)])
       (values key (map cdr group)))))
 
 (define (same-sign? a b)
   (or (and (bfpositive? a) (bfpositive? b))
       (and (bfnegative? a) (bfnegative? b))))
- 
+
+(define all-explanations
+  (list 'uflow-rescue
+        'u/u 'u/n 'o/o 'n/o
+        'o*u 'u*o 'n*u))
+
 (define (predicted-errors expr ctx pctx)
   (define cond-thres (bf 100))
+  (define maybe-cond-thres (bf 32))
+
   
-  (define subexprs (all-subexpressions expr))
-  (define spec-list (map prog->spec subexprs))
- 
+  (define subexprs
+    (all-subexpressions-rev expr (context-repr ctx)))
+  (define subexprs-list (map car subexprs))
```

Comment:
It would be nice to make `all-subexpressions-rev` return the same kind of thing as `all-subexpressions`, in other words have it return what is called `subexprs-list` here. Then the code could be a little nicer.

Reply:
Fine to do in another PR